packages/firestore/src/api/credentials.ts

```diff
@@ -306,6 +336,33 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
     );
     return new User(currentUid);
   }
+
+  /**
+   * Blocks the AsyncQueue until the next user is available. This is invoked
+   * on SDK start to wait for the first user token (or `null` if Auth is not yet
+   * loaded). If Auth is loaded after Firestore,
+   * `awaitTokenAndRaiseInitialEvent()` is also used to block Firestore until
+   * Auth is fully initialized.
+   *
+   * This function also invokes the change listener synchronously once a token
+   * is available.
+   */
+  private awaitTokenAndRaiseInitialEvent(): void {
+    this.invokeChangeListener = false; // Prevent double-firing of the listener
+    if (this.asyncQueue) {
+      // Create a new deferred Promise that gets resolved when we receive the
+      // next token. Ensure that all previous Promises also get resolved.
+      const awaitToken = new Deferred<void>();
+      void awaitToken.promise.then(() => awaitToken.resolve());
```

Comment:
`then` only fires if `awaitToken` resolves, why do we resolve it again in `then`? The comment suggests it ensures all previous Promises also get resolved? But I don't get how that works.

Reply:
This was meant to resolve the existing this.receivedUser Promise if another one was created later on. I changed how the listener gets called if auth is not used, and am no longer using the Promise twice. This allowed me to remove this logic.